python/phases.py

```diff
@@ -0,0 +1,107 @@
+from typing import Union, Set, Tuple
+import pathlib
+import os
+import copy
+import sys
+import time
+
+import lxml.html
+
+from traverse import traverse_and_replace
+
+PYTHON_PATH = pathlib.Path(__file__).parent.parent.resolve()
+CORE_ELEMENTS_PATH = (PYTHON_PATH / "elements").resolve()
+SAVED_PATH = copy.copy(sys.path)
+
+
+# def resolve_element(element_name: str, context: dict) -> dict:
+#     if element_name in context["course_elements"]:
+#         return context["course_elements"][element_name]
+#     elif element_name in context["c_elements"]:
+
+
+def render(data: dict, context: dict) -> Tuple[str, Set[str]]:
```

Comment:
Data and context should use TypedDicts for their annotations, which will make the comments below unnecessary.